src/libsystemd-network/dhcp6-option.c

```diff
@@ -370,47 +370,42 @@ int dhcp6_option_append_vendor_class(uint8_t **buf, size_t *buflen, char * const
         return dhcp6_option_append(buf, buflen, SD_DHCP6_OPTION_VENDOR_CLASS, total, p);
 }
 
-static int option_parse_hdr(uint8_t **buf, size_t *buflen, uint16_t *optcode, size_t *optlen) {
-        DHCP6Option *option = (DHCP6Option*) *buf;
-        uint16_t len;
-
-        assert_return(buf, -EINVAL);
-        assert_return(optcode, -EINVAL);
-        assert_return(optlen, -EINVAL);
-
-        if (*buflen < offsetof(DHCP6Option, data))
-                return -ENOMSG;
+int dhcp6_option_parse(
+                const uint8_t *buf,
+                size_t buflen,
+                size_t *offset,
+                uint16_t *ret_option_code,
+                size_t *ret_option_data_len,
+                const uint8_t **ret_option_data) {
 
-        len = be16toh(option->len);
+        const DHCP6Option *option;
+        size_t len;
 
-        if (len > *buflen)
-                return -ENOMSG;
+        assert(buf);
+        assert(offset);
+        assert(ret_option_code);
+        assert(ret_option_data_len);
+        assert(ret_option_data);
 
-        *optcode = be16toh(option->code);
-        *optlen = len;
+        if (buflen < offsetof(DHCP6Option, data))
+                return -EBADMSG;
 
-        *buf += 4;
-        *buflen -= 4;
+        if (*offset >= buflen)
+                return -EBADMSG;
 
-        return 0;
-}
+        if (*offset + offsetof(DHCP6Option, data) >= buflen)
+                return -EBADMSG;
 
-int dhcp6_option_parse(uint8_t **buf, size_t *buflen, uint16_t *optcode,
-                       size_t *optlen, uint8_t **optvalue) {
-        int r;
-
-        assert_return(buf && buflen && optcode && optlen && optvalue, -EINVAL);
-
-        r = option_parse_hdr(buf, buflen, optcode, optlen);
-        if (r < 0)
-                return r;
+        option = (const DHCP6Option*) (buf + *offset);
+        len = be16toh(option->len);
 
-        if (*optlen > *buflen)
-                return -ENOBUFS;
+        if (*offset + offsetof(DHCP6Option, data) + len > buflen)
```

Comment:
following the pattern above, maybe:

```c
if (len > buflen - offsetof(DHCP6Option, data) - *offset)
```
Because the rhs can't underflow now, because you check that already